### DISCUSSION TITLE
Mix FTXUI logger and spdlog lib

### DISCUSSION THREAD

**Discussion by Chasikanaft** [ArthurSonzogni/FTXUI#427] [Ideas]
Anyone did something like this? I saw an example FTXUI with _Windowslogger_, but I would like to format the messages as needed right away. Is it possible to override the "_logger class_" **spdlog** for output in **FTXUI**?

**Comment by Root3287** [comment]
Currently researching into this. What I think you could do is to add an `std::ostream` sink to your spdlog object. Then from there, parse out the output where it has `\n` and create a `ftxui::text()` per line. Then add all of those to a `ftxui::vbox()`. The issue is splitting at `\n`. So what you can do is make your own `std::ostream` sink. So far that is my theory, but I will come back and update this thread when I have working code.

**Reply by Root3287** [reply]
Well I got it done faster than I though...

<details>
<summary>ftxui_sink.hpp</summary>

```c++
#pragma once

#include <mutex>
#include <ostream>

#include <ftxui/dom/elements.hpp>

#include <spdlog/sinks/base_sink.h>

template<typename Mutex>
class ftxui_sink : public spdlog::sinks::base_sink<Mutex>{
public:
	inline std::vector<ftxui::Element> getLog() { return _log; };
protected:
	void sink_it_(const spdlog::details::log_msg& msg) override{
		spdlog::memory_buf_t formatted;
		spdlog::sinks::base_sink<Mutex>::formatter_->format(msg, formatted);

		std::stringstream ss;
		ss.write(formatted.data(), static_cast<std::streamsize>(formatted.size()));
		auto text = ftxui::text(ss.str());
		_log.push_back(text);
	};
	void flush_() override {};

	std::vector<ftxui::Element> _log;
};

using ftxui_sink_mt = ftxui_sink<std::mutex>;
using ftxui_sink_st = ftxui_sink<spdlog::details::null_mutex>;
```
</details>

<details>
<summary>main.cpp</summary>

```c++
_LOGGER = std::make_shared<ftxui_sink_mt>();
// Setup your logger how you want
// ...
// FTXUI code

auto log = ftxui::vbox(_LOGGER->getLog());
ftxui::Render(screen, log);
```
</details>
